trieste/models/gpflow/architectures.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2021 The Trieste Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" This file contains builders for some GPflow models supported in Trieste. """
+
+from __future__ import annotations
+
+from typing import Optional
+
+import gpflow
+import tensorflow as tf
+import tensorflow_probability as tfp
+from gpflow.models import GPR, SGPR, SVGP, VGP
+
+from ...data import Dataset
+from ...space import Box
+
+
+def build_gpr(data: Dataset, trainable_likelihood: bool = False, kernel_priors: bool = True) -> GPR:
+    """
+    Constructs a :class:`~gpflow.models.GPR` model with sensible initial parameters and
+    (optionally) priors. We use :class:`~gpflow.kernels.Matern52` kernel in the model as we found
+    it to be effective in most settings. We assume inputs are normalised to the unit hypercube for
+    setting sensible initial parameters for the kernel.
+
+    We set priors for kernel hyperparameters by default in order to stabilize model fitting. We
+    found the priors below to be highly effective for objective functions defined over the unit
+    hypercube. For objective functions with different scaling, other priors will likely be more
+    appropriate. Using priors allows for using maximum a posteriori estimate of these kernel
+    parameters during model fitting.
+
+    :param data: Dataset from the initial design, used for estimating the variance of observations.
+    :param trainable_likelihood: If set to `True` likelihood paramater is set to
+        trainable. By default set to `False`.
+    :param priors: If set to `True` (default) priors are set for kernel parameters.
+    :return: A :class:`~gpflow.models.GPR` model.
+    """
+    empirical_variance = tf.math.reduce_variance(data.observations)
+    dim = data.query_points.shape[-1]
+
+    kernel = gpflow.kernels.Matern52(variance=empirical_variance, lengthscales=[0.2] * dim)
```

Comment:
for consistency with SGPR, I'd vote for the latter solution

Reply:
we either assume we have a unit hypercube or try to be more general here, if we opt for the general approach I would also opt for search space as an additional input since it should always be available 

how to scale priors wrt the search space is less clear to me - if we cannot do that in a nice way, perhaps we should return to requiring unit hypercube and leave the user to do the normalisation, or we just say priors are unlikely to work well for search spaces that are not unit cube and should not be used

Reply:
I think assuming the unit hypercube is too much of a constraint (even though it's good practice). 

I've tried various things and all in all the scaling factor of the prior doesn't matter much, we can always use 1. And the above formula works well. 

 So I'd vote for using the search space. For COCO I actually use a formula to increase the lengthscale default with the square root of the dimension.